### ISSUE TITLE
`<memory>`: `unique_ptr<Incomplete>{nullptr}` compiler error with `/std:c++latest`

### ISSUE THREAD

**Issue by bebuch** [microsoft/STL#4417]
```cpp
#include <memory>

struct A;

struct B {
    B();
    ~B();

    std::unique_ptr<A> ptr{nullptr};
};

int main(){
    B b;
}
```

This compiles in C++23 mode with GCC and Clang and in C++20 mode with MSVC.

[Compiler Explorer](https://godbolt.org/z/bKaYG3hqG)

It fails to compile in C++23 with MSVC. The [support told me](https://developercommunity.visualstudio.com/t/10514504) that this is the expected behavior. I don't think this is true. Note this is not the same as #4215 because I initialize with a `nullptr`. The `unique_ptr` destructor should only be instantiated in the definition of `~B`, which is in another code unit where `A` is complete.

I asked a language lawyer question on Stack Overflow, but got no answer yet. The comments also seem to indicate that the MSVC `/std:c++latest` behavior is wrong:

- https://stackoverflow.com/questions/78040562

**Comment by bebuch** [comment]
Great examples, thanks!

I assume that the standard does not define what of these must and must not be compilable?

Is it possible to implement `unique_ptr` with a C++23 compilant compiler so that all these examples can be compiled, even if it is `constexpr`? If so, I think it would make sense to implement it this way and submit a paper that sets this behavior as the standard per DR.

It would be useful if the compilers would accept or reject the code uniformly.

**Comment by StephanTLavavej** [comment]
> The [support told me](https://developercommunity.visualstudio.com/t/10514504) that this is the expected behavior.

It's still me here! But I'll ask the other STL maintainers during our weekly meeting next Wednesday, it's possible that I was wrong.

clang-cl accepts your original code with MSVC's STL, so this is really a question about the MSVC compiler's behavior. We may need to reduce this to a library-free example and send it to the compiler team.

**Comment by frederick-vs-ja** [comment]
Reduced version ([Godbolt link](https://godbolt.org/z/zP9b3MhEr)) which is only rejected by MSVC. If either `constexpr` or default member initializer is not used, MSVC also accepts this.

```C++
template<class T>
struct holder {
    holder() = default;

    holder(const holder&) = delete;
    holder& operator=(const holder&) = delete;

#ifdef CONSTEXPR_DTOR
    constexpr
#endif
    ~holder()
    {
        static_assert(sizeof(T) || true);
    }
};

struct A;

struct B {
    B();
    ~B();

    holder<A> p
#ifdef USING_DMI
    {}
#endif
    ;
};

int main()
{
    B b [[maybe_unused]];
}
```

It seems that MSVC the `constexpr` on destructor and the default member initializer cause _something_ to instantiate the destructor of `holder<A>`.
However, given the templated version compiles fine ([Godbolt link](https://godbolt.org/z/778bnz1Gn)), I have no idea on what such "something" is. Perhaps MSVC is buggy.

<details><summary>Templated version</summary>
<p>

```C++
template<class T>
struct holder {
    holder() = default;

    holder(const holder&) = delete;
    holder& operator=(const holder&) = delete;

    constexpr ~holder()
    {
        static_assert(sizeof(T) || true);
    }
};

struct A;

template<class = void>
struct B {
    B();
    ~B();

    holder<A> p{};
};

using BB = B<>;

int main()
{
    BB bb [[maybe_unused]];
}
```

</p>
</details>

> Every C++ standard library triggers a `static_assert` about incomplete type

This is [[unique.ptr.dltr.dflt]/3](https://eel.is/c++draft/unique.ptr.dltr.dflt#3).

**Comment by bebuch** [comment]
```cpp
holder<A> a{};             // accepted by GCC and clang, rejected by MSVC
holder<A> b = {};          // accepted by clang, rejected by GCC and MSVC
holder<A> c = holder<A>{}; // always rejected
```

I think that case `a` and `b` should both compile, because both are [List-initialization syntax](https://en.cppreference.com/w/cpp/language/initialization).

Case `c` is rejected correctly by all compiles, because it creates a temporary that must be destructed.

I filed a bug report to GCC:

- https://gcc.gnu.org/bugzilla/show_bug.cgi?id=114076

**Comment by StephanTLavavej** [comment]
I've reported this compiler bug as DevCom-10604135 "MSVC instantiates constexpr destructors too early when data member initializers are involved". Thanks @frederick-vs-ja for the fully reduced repro. (One note - it's best for repros to demonstrate the bug without needing any additional macros to be defined; if you want to present modes that work, they should be activated by any macros.)

Since we have no workaround for the STL, I'm going to close this issue. Thanks again for bringing it to our attention and objecting when we tried to close it too early :smile_cat:
